### PR TITLE
AWS fargate collector

### PR DESCRIPTION
**Pull request by yogeshprasad**
AWS Fargate collector allows to collect the Fargate task metrics and sends to the Wavefront.

### CODE REVIEW

**Comment by vikramraman** [comment]
What's changed compared to the cadvisor storage driver?

**Comment by yogeshprasad** [comment]
Cadvisor collects the container information using the docker service.
Fargate does not allow any communication from docker service it means we can not collect the data using cadvisor.

wavefront fargate collector uses two APIs 169.254.170.2/v2/metadata and 169.254.170.2/v2/stats to collect the data.
These APIs are exposed by fargate.

https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v2.html

File: adapter.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	// Waitgroup to keep track of goroutines
+	var wg sync.WaitGroup
+
+	// Map to support the list of storage driver
+	funcs := map[string]func(map[string]string, *sync.WaitGroup){"wavefront": storage.Wavefront}
+
+	// Do not use the storage driver if user is not intended
+	if *storageDriver == "" {
+		log.Printf("Storage driver is not supplied")
```
Comment (vikramraman) [inline adapter.go]: Should we exit here? You can just do `log.Fatal("Storage driver is not supplied")`

File: adapter.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	} else {
+		// Check is supplied storage driver is supported
+		_, has := funcs[*storageDriver]
+		if !has {
+			log.Printf("Supplied storage driver is not supported")
```
Comment (vikramraman) [inline adapter.go]: Ditto. `log.Fatal(...)` instead of print and return.

File: adapter.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		// Process the storageType specific inputs
+		storageDriverOpt := map[string]string{}
+		if *storageDriverOptStr != "" {
+			for _, item := range strings.Split(*storageDriverOptStr, " ") {
+				kwargs := strings.Split(item, "=")
+				storageDriverOpt[strings.TrimSpace(kwargs[0])] = strings.TrimSpace(kwargs[1])
+			}
+		}
+
+		wg.Add(1) // create a waightgroup entry for a goroutine
```
Comment (vikramraman) [inline adapter.go]: nitpick: typo in wait group.

File: adapter.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+		// Call the storage driver
+		go funcs[*storageDriver](storageDriverOpt, &wg)
+	}
+
+	// Start the server if port is supplied
+	if *port != 0 {
+
```
Comment (vikramraman) [inline adapter.go]: remove blank line

File: adapter.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+		http.Handle("/", http.FileServer(http.Dir("./static")))
+		http.HandleFunc("/metrics", handler.MetricsHandler)
+		http.HandleFunc("/stats", handler.StatsHandler)
+		http.HandleFunc("/metadata", handler.MetadataHandler)
+
+		log.Printf("Server is listening on port " + strconv.Itoa(*port))
+		http.ListenAndServe(":"+strconv.Itoa(*port), nil)
+	} else {
+		log.Printf("Listener configurationn is not supplied")
```
Comment (vikramraman) [inline adapter.go]: ditto. Fatal?

File: storage/wavefront.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import (
+	"fmt"
+	"log"
+	"os"
+	"strconv"
+	"sync"
+	"time"
+
+	wavefrontSenders "github.com/wavefronthq/wavefront-sdk-go/senders"
+	. "github.com/yogeshprasad/fargate/backend"
```
Comment (vikramraman) [inline storage/wavefront.go]: Change this?

File: static/index.html [hunk 7]
```diff
@@ -0,0 +1,72 @@
+<!DOCTYPE html>
```
Comment (vikramraman) [inline static/index.html:1]: remove this file?

File: storage/wavefront.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+			MaxBufferSize: 50000,
+
+			// interval (in seconds) at which to flush data to Wavefront. defaults to 1 Second.
+			// together with batch size controls the max theoretical throughput of the sender.
+			FlushIntervalSeconds: metricFlushInterval,
+		}
+
+		sender, err = wavefrontSenders.NewDirectSender(directCfg)
+		if err != nil {
+			log.Println(err.Error())
```
Comment (vikramraman) [inline storage/wavefront.go]: Just do a `log.Fatal` if the configuration if we cannot create either sender.

File: storage/wavefront.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+			break
+		}
+		if metrics == nil {
+			log.Println("Data not found")
+		}
+		hostName, _ := os.Hostname()
+		for _, item := range metrics {
+			sender.SendMetric(metricPrefix+item.Name, item.Value, 0, hostName, item.Tags)
+		}
+		time.Sleep(2 * time.Second)
```
Comment (vikramraman) [inline storage/wavefront.go]: We should make this interval configurable. And probably default to once a minute? Collecting every 2 seconds seems expensive and will generate a lot of PPS for end users.

Comment (vikramraman) [reply storage/wavefront.go]: Also use a ticker here instead of sleep.

File: storage/wavefront.go [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+			wg.Done()
+			return
+		}
+	}
+
+	for {
+		metrics, err := GetMetrics()
+		if err != nil {
+			log.Println(err.Error())
+			break
```
Comment (vikramraman) [inline storage/wavefront.go]: Don't think we should exit if there is an error on collection. There may be intermittent errors? We should just log and continue the next time around. Would also be useful if we emitted some health metrics into Wavefront once a minute about the health of the collector itself.

**Review by vikramraman — CHANGES_REQUESTED** [review]
@yogeshprasad Left a few comments. Please address them.

**Comment by vikramraman** [comment]
Also include your dependencies using go modules.

File: adapter.go (hunk 1 above)
Comment (yogeshprasad) [reply adapter.go]: as of now, it is optional, so a user can debug the collected metrics without supplying the storage driver.

File: static/index.html (hunk 7 above)
Comment (yogeshprasad) [reply static/index.html:1]: I have enabled the web UI as well to debug the collected metrics.

File: adapter.go (hunk 1 above)
Comment (vikramraman) [reply adapter.go]: Can we just add a `--test` or `--debug` flag or such for that. So we'll always quit if driver is not supplied. But if debug flag is provided we will only log the metrics etc.

File: storage/wavefront.go (hunk 10 above)
Comment (yogeshprasad) [reply storage/wavefront.go]: We are sending the collector statistics as well

**Review by vikramraman — APPROVED** [review]

**Comment by vikramraman** [comment]
@yogeshprasad lgtm. will merge this PR. But we should add unit tests for this and add a `.travis.yml` file for continuous builds.

**Pull request merged**